File: DAYS/Day62-Same Binary Tree/sameTree.py
# Binary tree - Same Tree (LeetCode 100)
# Given the roots of two binary trees p and q, write a function to check if they are the same or not.
# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
# Example 1:Input: p = [1,2,3], q = [1,2,3] Output: true
# Example 2:Input: p = [1,2], q = [1,null,2] Output: false
# Different Approaches: 
# 1. Recursive DFS-Time O(n), Space O(n)
# 2. Iterative BFS-Time O(n), Space O(n)
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sameTree(p,q):
        if not p and not q: return True
        if not p or not q or p.val != q.val: return False
        return Solution.sameTree(p.left, q.left) and Solution.sameTree(p.right, q.right)

# Test Cases
if __name__ == "__main__":
    # Example 1
    p1 = TreeNode(1)
    p1.left = TreeNode(2)
    p1.right = TreeNode(3)
    q1 = TreeNode(1)
    q1.left = TreeNode(2)
    q1.right = TreeNode(3)
    print(Solution.sameTree(p1, q1))  # Output: True

    # Example 2
    p2 = TreeNode(1)
    p2.left = TreeNode(2)
    q2 = TreeNode(1)
    q2.right = TreeNode(2)
    print(Solution.sameTree(p2, q2))  # Output: False
# Complexity Analysis:
# Time Complexity: O(n)
# Space Complexity: O(n)

File: DAYS/Day62-Same Binary Tree/sameTree.js
// Binary Tree- Same Tree (LeetCode 100)
// Given the roots of two binary trees p and q, write a function to check if they are the same or not.
// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
// Example 1: Input: p = [1,2,3], q = [1,2,3] Output: true
// Example 2: Input: p = [1,2], q = [1,null,2] Output: false
// Example 3: Input: p = [1,2,1], q = [1,1,2] Output: false

// Different Approaches:
// 1. Recursive Approach - Time O(n), Space O(n)
// 2. Iterative Approach using BFS - Time O(n), Space O(n)

// Leetcode link: https://leetcode.com/problems/same-tree/
// Solution:
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
function sameTree(p,q){
    if(!p && !q) return true;
    if(!p || !q || p.val !== q.val) return false;
    return sameTree(p.left, q.left) && sameTree(p.right, q.right);
}
// Test Cases
const tree1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
const tree2 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
console.log(sameTree(tree1, tree2)); // true
const tree3 = new TreeNode(1, new TreeNode(2), null);
const tree4 = new TreeNode(1, null, new TreeNode(2));
console.log(sameTree(tree3, tree4)); // false
const tree5 = new TreeNode(1, new TreeNode(2), new TreeNode(1));
const tree6 = new TreeNode(1, new TreeNode(1), new TreeNode(2));
console.log(sameTree(tree5, tree6)); // false

// Both Approches Complexity Analysis:Time Complexity: O(n) , Space Complexity: O(n)